cpp/include/raft/neighbors/detail/cagra/compute_distance_vpq.cuh

```diff
@@ -56,9 +56,17 @@ struct cagra_q_dataset_descriptor_t : public dataset_descriptor_base_t<half, DIS
     if constexpr (std::is_same<CODE_BOOK_T, half>::value) {
       for (unsigned i = threadIdx.x * 2; i < (1 << PQ_BITS) * PQ_LEN; i += blockDim.x * 2) {
         half2 buf2;
-        buf2.x                                                   = pq_code_book_ptr[i];
-        buf2.y                                                   = pq_code_book_ptr[i + 1];
-        (reinterpret_cast<half2*>(smem_pq_code_book_ptr + i))[0] = buf2;
+        buf2.x = pq_code_book_ptr[i];
+        buf2.y = pq_code_book_ptr[i + 1];
+
+        // Change the order of PQ code book array to reduce the
+        // frequency of bank conflicts.
+        constexpr auto num_elements_per_bank  = 4 / utils::size_of<CODE_BOOK_T>();
+        constexpr auto num_banks_per_subspace = PQ_LEN / num_elements_per_bank;
+        const auto j                          = i / num_elements_per_bank;
+        const auto smem_index =
+          (j / num_banks_per_subspace) + (j % num_banks_per_subspace) * (1 << PQ_BITS);
+        reinterpret_cast<half2*>(smem_pq_code_book_ptr)[smem_index] = buf2;
```

Comment:
Do we need a similar adjustment to the logic in the other branch of the if (CODE_BOOK_T is float)?

Reply:
No, we don't need it because the current `compute_similarity` only supports `half` as the `CODE_BOOK_T`. I removed the branch of the CODE_BOOK_T == float and added a static_assert to check the `CODE_BOOK_T` type. If it is better to support `float` as the `CODE_BOOK_T` type, I'll create a PR for that.

Reply:
Sounds good, thanks!